jdbc-driver/src/main/java/com/google/cloud/dataproc/jdbc/DataprocInfo.java

```diff
@@ -0,0 +1,169 @@
+            return cluster.getClusterName();
+        } catch (ApiException e) {
+            throw new InvalidURLException(
+                    String.format(
+                            "Unable to retrieve cluster information for %s/%s/%s.\n%s",
+                            params.projectId(), params.region(), params.clusterName(), e.getMessage()));
+        }
+    }
+
+    /**
+     * Formats the user passed in clusterPoolLabel value to the filter format that Dataproc API
+     * accpets: field = value [AND [field = value]] ...
+     *
+     * @return formatted filter string accepted by Dataproc Client API
+     */
+    public String formatClusterFilterString() throws InvalidURLException {
+        Map<String, List<String>> labels = new LinkedHashMap<>();
+        labels.put("status.state", new ArrayList<>(Arrays.asList("ACTIVE")));
+
+        if (params.clusterPoolLabel() != null) {
+            for (String label : params.clusterPoolLabel().split(":")) {
+                checkUrl(
+                        !label.isEmpty() && label.contains("=") && !label.startsWith("status"),
+                        "'%s' Invalid clusterPoolLabel.\n"
+                                + "Please do not provide cluster status as label.\n"
+                                + "Example format: clusterPoolLabel=field1=value1:field2=value2'",
+                        label);
+                String field = label.split("=")[0];
+                String value = label.split("=")[1];
+                // filed can be label.[key] or clusterName
+                labels
+                        .computeIfAbsent(
+                                field.equals("clusterName") ? field : "labels." + field, k -> new ArrayList<>())
+                        .add(value);
+            }
+        }
+
+        String formattedFilter =
+                labels.entrySet().stream()
+                        .flatMap(entry -> entry.getValue().stream().map(v -> entry.getKey() + " = " + v))
+                        .collect(Collectors.joining(" AND "));
+        return formattedFilter;
+    }
+
+    /**
+     * Supports picking a cluster from cluster pool.
+     *
+     * @param filter formatted filter that matches the conditions client passed in
+     * @return name of the suitable cluster with cluster pool
+     * @throws IOException
+     */
+    public String findClusterInPool(String filter) throws InvalidURLException {
+        try {
+            String clusterName = null;
+            for (Cluster response :
+                    clusterControllerClient
+                            .listClusters(params.projectId(), params.region(), filter)
+                            .iterateAll()) {
+                // TODO: decide on how to pick a cluster among cluster pool
+                clusterName = response.getClusterName();
+            }
+            clusterControllerClient.close();
+            return clusterName;
+        } catch (ApiException e) {
+            throw new InvalidURLException(
```

Comment:
Here and elsewhere, you can instead create a constructor that delegates to this one: https://docs.oracle.com/javase/7/docs/api/java/sql/SQLException.html#SQLException(java.lang.String,%20java.lang.Throwable)

That way, super.getMessage() will add the underlying exception message into the message for you.